flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/TableUtil.java

```diff
@@ -0,0 +1,432 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.utils;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava18.com.google.common.base.Joiner;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Utility class for the operations on Table, such as the operations on column name, the operations on column type and
+ * so on.
+ */
+public class TableUtil {
+	/**
+	 * Return a lower temp table name in lower case and use "-" as connector character.
+	 *
+	 * @return tableName
+	 */
+	public static synchronized String getTempTableName() {
+		return ("temp_" + UUID.randomUUID().toString().replaceAll("-", "_"))
+			.toLowerCase();
+	}
+
+	/**
+	 * Find the index of <code>targetCol</code> in string array <code>tableCols</code>. It will ignore the case of the
+	 * tableCols.
+	 *
+	 * @param tableCols a string array among which to find the targetCol.
+	 * @param targetCol the targetCol to find.
+	 * @return the index of the targetCol, if not found, returns -1.
+	 */
+	public static int findColIndex(String[] tableCols, String targetCol) {
+		Preconditions.checkNotNull(targetCol, "targetCol is null!");
+		for (int i = 0; i < tableCols.length; i++) {
+			if (targetCol.equalsIgnoreCase(tableCols[i])) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	/**
+	 * Find the indices of <code>targetCols</code> in string array <code>tableCols</code>. If
+	 * <code>targetCols</code> is
+	 * null, it will be replaced by the <code>tableCols</code>
+	 *
+	 * @param tableCols  a string array among which to find the targetCols.
+	 * @param targetCols the targetCols to find.
+	 * @return the indices of the targetCols.
+	 */
+	public static int[] findColIndices(String[] tableCols, String[] targetCols) {
+		if (targetCols == null) {
+			int[] indices = new int[tableCols.length];
+			for (int i = 0; i < tableCols.length; i++) {
+				indices[i] = i;
+			}
+			return indices;
+		}
+		int[] indices = new int[targetCols.length];
+		for (int i = 0; i < indices.length; i++) {
```

Comment:
one more question is: would this function be used (1) only once (e.g. during table program compile time) or (2) per-row (e.g. as part of the job graph and execute during runtime) ? 

if (2), there's ways to avoid the double loop for performance improvement but if (1) is true there's no need IMO.

Reply:
Thanks, this function should be called at the flink client. so it belongs to (1)

Reply:
Thanks, this function should be called at the flink client. so it belongs to (1)

Reply:
thanks for the explanation. yes it looks good to me